core/src/main/scala/org/apache/spark/storage/DiskBlockManager.scala

```diff
@@ -53,35 +53,98 @@ private[spark] class DiskBlockManager(blockManager: BlockManager, conf: SparkCon
 
   private val shutdownHook = addShutdownHook()
 
+  private abstract class FileAllocationStrategy {
+    def apply(filename: String): File
+
+    protected def getFile(filename: String, storageDirs: Array[File]): File = {
+      require(storageDirs.nonEmpty, "could not find file when the directories are empty")
+
+      // Figure out which local directory it hashes to, and which subdirectory in that
+      val hash = Utils.nonNegativeHash(filename)
+      val dirId = localDirs.indexOf(storageDirs(hash % storageDirs.length))
+      val subDirId = (hash / storageDirs.length) % subDirsPerLocalDir
+
+      // Create the subdirectory if it doesn't already exist
+      val subDir = subDirs(dirId).synchronized {
+        val old = subDirs(dirId)(subDirId)
+        if (old != null) {
+          old
+        } else {
+          val newDir = new File(localDirs(dirId), "%02x".format(subDirId))
+          if (!newDir.exists() && !newDir.mkdir()) {
+            throw new IOException(s"Failed to create local dir in $newDir.")
+          }
+          subDirs(dirId)(subDirId) = newDir
+          newDir
+        }
+      }
+
+      new File(subDir, filename)
+    }
+  }
+
   /** Looks up a file by hashing it into one of our local subdirectories. */
   // This method should be kept in sync with
   // org.apache.spark.network.shuffle.ExternalShuffleBlockResolver#getFile().
```

Comment:
We need to sync up the `org.apache.spark.network.shuffle.ExternalShuffleBlockResolver#getFile()`, so this feature can also be used within the external shuffle service in dynamic allocation of yarn.